### PR TITLE
Integration with ChatGLM

### PR DESCRIPTION
**Pull request by Martin7-1**
support [chatglm](https://github.com/THUDM/ChatGLM-6B), which was mentioned in #267 .

[chatglm2](https://github.com/THUDM/ChatGLM2-6B) and [chatglm3](https://github.com/THUDM/ChatGLM3) api are compatible with openai, so It is enough to support chatglm.

because chatglm does not have official docker image, so I don't know how to use `Testcontainers` to do test. (I'm not familiar with `Testcontainers`, so for now I have to copy test from the other modules, lol). The test will update using `Testcontainers` after I learn about it in few days.

### CODE REVIEW

File: langchain4j-chatglm/src/main/java/dev/langchain4j/model/chatglm/ChatCompletionRequest.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.util.List;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
+public class ChatCompletionRequest {
```
Comment (langchain4j) [inline langchain4j-chatglm/src/main/java/dev/langchain4j/model/chatglm/ChatCompletionRequest.java]: please make all classes except ChatGLMChatModel package-private

File: langchain4j-chatglm/src/main/java/dev/langchain4j/model/chatglm/ChatGLMChatModel.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static dev.langchain4j.internal.RetryUtils.withRetry;
+import static dev.langchain4j.internal.Utils.getOrDefault;
+
+public class ChatGLMChatModel implements ChatLanguageModel {
```
Comment (langchain4j) [inline langchain4j-chatglm/src/main/java/dev/langchain4j/model/chatglm/ChatGLMChatModel.java]: please rename `...GLM...` into `...Glm...` in class names everywhere

File: langchain4j-chatglm/src/main/java/dev/langchain4j/model/chatglm/ChatGLMChatModel.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        this.client = new ChatGLMClient(baseUrl, timeout);
+        this.temperature = getOrDefault(temperature, 0.7);
+        this.maxRetries = getOrDefault(maxRetries, 3);
+        this.topP = topP;
+        this.maxLength = maxLength;
+    }
+
+
+    @Override
+    public Response<AiMessage> generate(List<ChatMessage> messages) {
```
Comment (langchain4j) [inline langchain4j-chatglm/src/main/java/dev/langchain4j/model/chatglm/ChatGLMChatModel.java]: Please validate explicitly that no `SystemMessage` is present in `messages` (if it is not supported by ChatGLM)

File: langchain4j-chatglm/src/main/java/dev/langchain4j/model/chatglm/ChatGLMChatModel.java (hunk 2 above)
Comment (langchain4j) [inline langchain4j-chatglm/src/main/java/dev/langchain4j/model/chatglm/ChatGLMChatModel.java]: Please add a javadoc explaining for which version of ChatGLM this class is usable

**Review by ghost — CHANGES_REQUESTED** [review]
@Martin7-1 thank you for your contribution! Apart from some minor comments in the review, I am wondering if it makes sense to support old version of ChatGLM if newer versions support OpenAI API? 
image
Is old version widely used?

**Comment by Martin7-1** [comment]
> @Martin7-1 thank you for your contribution! Apart from some minor comments in the review, I am wondering if it makes sense to support old version of ChatGLM if newer versions support OpenAI API? image Is old version widely used?

There are two reasons to support `ChatGLM` old version:

* `ChatGLM` has most github stars, Although ChatGLM2 and ChatGLM3 are newer, ChatGLM is still used by many developers.
* `langchain` also support `ChatGLM`, I think it gives our users more models to choose from. (see [langchain](https://github.com/langchain-ai/langchain/blob/master/libs/community/langchain_community/llms/chatglm.py) and [langchain doc](https://python.langchain.com/docs/integrations/llms/chatglm))

File: langchain4j-chatglm/src/main/java/dev/langchain4j/model/chatglm/ChatCompletionRequest.java (hunk 1 above)
Comment (Martin7-1) [reply langchain4j-chatglm/src/main/java/dev/langchain4j/model/chatglm/ChatCompletionRequest.java]: done

File: langchain4j-chatglm/src/main/java/dev/langchain4j/model/chatglm/ChatGLMChatModel.java (hunk 2 above)
Comment (Martin7-1) [reply langchain4j-chatglm/src/main/java/dev/langchain4j/model/chatglm/ChatGLMChatModel.java]: done

File: langchain4j-chatglm/src/main/java/dev/langchain4j/model/chatglm/ChatGLMChatModel.java (hunk 3 above)
Comment (Martin7-1) [reply langchain4j-chatglm/src/main/java/dev/langchain4j/model/chatglm/ChatGLMChatModel.java]: done

File: langchain4j-chatglm/src/main/java/dev/langchain4j/model/chatglm/ChatGLMChatModel.java (hunk 2 above)
Comment (Martin7-1) [reply langchain4j-chatglm/src/main/java/dev/langchain4j/model/chatglm/ChatGLMChatModel.java]: done

**Comment by langchain4j** [comment]
@Martin7-1 sure, no problem, let's support an older version. But I hope that you will help maintaining this module :)

**Comment by Martin7-1** [comment]
> @Martin7-1 sure, no problem, let's support an older version. But I hope that you will help maintaining this module :)

No problem. It's my pleasure.

**Review by ghost — APPROVED** [review]
@Martin7-1 thank you, good job!

**Comment by pigeon2049** [comment]
Hi, can you add some docs of integrate with local ChatGLM3?
https://github.com/langchain4j/langchain4j-examples/

**Comment by Martin7-1** [comment]
> Hi, can you add some docs of integrate with local ChatGLM3? https://github.com/langchain4j/langchain4j-examples/

@pigeon2049 Hi, ChatGLM3's api is compatible with openai api, so just use `langchain4j-open-ai`

**Pull request merged**